compiler/modules/types/xml.bal

```diff
@@ -0,0 +1,196 @@
+// Implementation specific to basic type xml.
+
+public type XmlSubtype readonly & record {|
+    // Each xml singleton and the empty sequence is considered a primitive.
+    int primitives;
+    Bdd sequence;
+|};
+
+// Each bit represents a XML singleton subtype or a empty sequence.
+// `primitives` fields are composed of allowed values from this singleton set.
+// Atom of the `Bdd` represented by `sequence` field is also composed of these to indicate 
+// non-empty sequence of allowed singletons.
+const int XML_PRIMITIVE_NEVER      = 1;
+const int XML_PRIMITIVE_TEXT       = 1 << 1;
+const int XML_PRIMITIVE_ELEMENT_RO = 1 << 2;
+const int XML_PRIMITIVE_PI_RO      = 1 << 3;
+const int XML_PRIMITIVE_COMMENT_RO = 1 << 4;
+const int XML_PRIMITIVE_ELEMENT_RW = 1 << 5;
+const int XML_PRIMITIVE_PI_RW      = 1 << 6;
+const int XML_PRIMITIVE_COMMENT_RW = 1 << 7;
+
+const int XML_PRIMITIVE_RO_SINGLETONS = XML_PRIMITIVE_TEXT | XML_PRIMITIVE_ELEMENT_RO | XML_PRIMITIVE_PI_RO | XML_PRIMITIVE_COMMENT_RO;
+const int XML_PRIMITIVE_RO_MASK = XML_PRIMITIVE_NEVER | XML_PRIMITIVE_RO_SINGLETONS;
+const int XML_PRIMITIVE_RW_MASK = XML_PRIMITIVE_ELEMENT_RW | XML_PRIMITIVE_PI_RW | XML_PRIMITIVE_COMMENT_RW;
+
+final XmlSubtype xmlRoTop = { primitives: XML_PRIMITIVE_RO_MASK, sequence: bddAtom(XML_PRIMITIVE_RO_SINGLETONS) };
+final XmlSubtype xmlRwTop = { primitives: XML_PRIMITIVE_RW_MASK, sequence: bddAtom(XML_PRIMITIVE_RW_MASK|XML_PRIMITIVE_RO_SINGLETONS) };
+
+function xmlSingleton(int primitives) returns SemType {
+    return createXmlSemtype(
+        createXmlSubtype(true, primitives, false), 
+        createXmlSubtype(false, primitives, false)
+    );
+}
+
+public function xmlSequence(SemType constituentType) returns SemType {
+    if constituentType == NEVER {
+        return xmlSequence(xmlSingleton(XML_PRIMITIVE_NEVER));
+    }
+    if constituentType is UniformTypeBitSet {
+        return constituentType;
+    }   
+    else {
+        SubtypeData ro = getComplexSubtypeData(constituentType, UT_XML_RO);
+        ro = ro is boolean ? ro : makeSequence(true, <XmlSubtype>ro);
+        
+        SubtypeData rw = getComplexSubtypeData(constituentType, UT_XML_RW);
+        rw = rw is boolean ? rw : makeSequence(false, <XmlSubtype>rw);
+        
+        return createXmlSemtype(ro, rw);
+    }
+}
+
+function makeSequence(boolean roPart, XmlSubtype d) returns XmlSubtype {
```

Comment:
How does this deal with the case where we are creating the rw part, and d.sequence is false and all the primitives are ro?

e.g. `xml<readonly & xml:Element> & !readonly` should equivalent to `never`.

Reply:
I think I forgot to update this when addressing https://github.com/ballerina-platform/nballerina/pull/562#pullrequestreview-780698845; will update.